--- 
name: wrapping-exceptions-zzzzz
layout: post
title: Wrapping Exceptions zzzZZ
time: 2010-04-25 18:19:00 +01:00
---
I'm totally bored of wrapping exceptions in Java, <br /><br /><pre class="java:nogutter:nocontrols" name="code">try {<br />   // do something<br />} catch (BoredomException e) {<br />   // do something else<br />} <br /></pre>It's verbose, ugly and has nothing to do with what you're really trying to convey. It's just noise (more or less). For example, when using the just <i>dreadful</i> Google Data API to access my calendar, I wrapped a couple of underlying Google services to be able to mock. Each service wanted to throw a bunch of Google specific exceptions which I, of course, wanted to rethrow as application specific. Boring...<br /><br /><pre class="java:nogutter:nocontrols" name="code">public O call() throws CalendarException {<br />    try {<br />        return service.call();  // the call to the google service<br />    } catch (MalformedURLException e) {<br />        throw new CalendarException(BAD_URL_MESSAGE, e); <br />    } catch (IOException e) {<br />        throw new CalendarException(IO_EXCEPTION_MESSAGE, e);<br />    } catch (AuthenticationException e) {<br />        throw new CalendarException(AUTHENTICATION_MESSAGE, e);<br />    } catch (ServiceException e) {<br />        throw new CalendarException(SERVICE_EXCEPTION_MESSAGE, e);<br />    }<br />}<br /></pre><br />If I didn't delegate like this, every service call would have to wrap and handle the google exceptions. As I've done this a few times, I decided to add it to <a href="http://code.google.com/p/tempus-fugit/">tempus-fugit</a> as a <code>ExceptionWrapper</code> class. Using this class, you can wrap a <code>Callable</code> to rethrow any caught exception as some other (including the underlying exception as the <code>cause</code>).<br /><br />So, in a similar way to the above, the client can ignore any declared exceptions and just rethrow them in-line. For example,<br /><br /><pre class="java:nogutter:nocontrols" name="code">wrapAnyException(new Callable&lt;Object&gt;() {     <br />    @Override     <br />    public Object call() throws ServiceException {<br />         // nasty code throwing a bunch of exceptions     <br />    } <br />}, with(CalendarException.class));<br /></pre>when this is in-lined further, it hopefully becomes more succinct.  <br /><pre class="java:nogutter:nocontrols" name="code">wrapAnyException(serviceCall(), with(CalendarException.class));</pre><br />This will wrap any exception and rethrow as a new <code>CalendarException</code> to include as the cause any underlying exception. It uses reflection to create the new exception, and forces the syntactically sugary <code>with</code> by taking a <code>WithException</code> as the second parameter.<br /><br />It's a candidate feature for tempus-fugit 1.1 for now, and will go in for sure if it gets some millage. Let me know if you find it useful. Source is <a href="http://code.google.com/p/tempus-fugit/source/browse/trunk/tempus-fugit/src/main/java/com/google/code/tempusfugit/ExceptionWrapper.java">here</a> with the test <a href="http://code.google.com/p/tempus-fugit/source/browse/trunk/tempus-fugit/src/test/java/com/google/code/tempusfugit/SimpleExceptionWrapperTest.java">here</a>.<br /><br /><div><img alt="" border="0" id="BLOGGER_PHOTO_ID_5283328307719135378" src="http://4.bp.blogspot.com/_-uMxV_fCbC4/SVInGoVdYJI/AAAAAAAAC08/I4RV1KzCyPo/s320/gibble_22x22.png" style="cursor: pointer; float: left; height: 22px; margin: 0pt 10px 10px 0pt; width: 22px;" /></div><div class="blogger-post-footer"><img width='1' height='1' src='https://blogger.googleusercontent.com/tracker/20485368-594689996510575973?l=pequenoperro.blogspot.com' alt='' /></div>
